ecommerce/processes/lib/processes/reservation_process.rb

```diff
@@ -28,10 +21,29 @@ def call(event)
     private
 
     def reserve_stock(state)
+      unavailable_products = []
       state.order_lines.each do |product_id, quantity|
         command_bus.(Inventory::Reserve.new(product_id: product_id, quantity: quantity))
         state.product_reserved(product_id)
+      rescue Inventory::InventoryEntry::InventoryNotAvailable
+        unavailable_products << product_id
+      end
+
+      if unavailable_products.empty?
+        event = ReservationProcessSuceeded.new(data: { order_id: state.order_id })
+      else
+        release_stock(state)
+        event = ReservationProcessFailed.new(data: { order_id: state.order_id, unavailable_products: unavailable_products })
       end
+      event_store.publish(event, stream_name: stream_name(state.order_id))
```

Comment:
ReservationProcess now emits events. We then subscribe to these events in the controller and display proper errors to the user.